### PR TITLE
New settings added to FIM synchronization, `response_timeout` and `max_interval`

### PR DESCRIPTION
**Pull request by jotacarma90**
|Related issue|
|---|
|https://github.com/wazuh/wazuh/issues/14069|

## Description
Hello team,
this PR is to implement the two new settings related to FIM synchronization:

- **`response_timeout`**: waiting time needed since a sync message is sent or received, until we allow to start the next sync.
- **`max_interval`**: On each failed synchronization the interval used between synchronizations is doubled. This max_interval represents the highest value that the interval can take using this system.

Added both settings to the configuration parser and configuration on demand.
We have also added the necessary algorithm so that consecutive synchronizations do not overlap.

## Algorithm operation

FIM synchronization works through a loop every `currentSyncInterval` seconds. This variable, in the beginning, corresponds to the interval configured in the synchronization section, although it can be modified if the synchronization fails (a failed synchronization is one in which the manager has responded with some message or the agent has had to send DB entries with messages of type `state`).

In each turn of the loop, the time since the last message received from the manager (or state sent by the agent) is calculated, if not enough time has elapsed (`m_syncResponseTimeout` seconds), the next synchronization is not started, and consequently, the `currentSyncInterval` is doubled. If enough time has elapsed, a new synchronization is launched, and if the previous one was successful, the interval is reset to its configured value.

https://github.com/wazuh/wazuh/blob/e8e28b94d61a49f422efd2994c8c4f807691f6e0/src/syscheckd/src/db/src/fimDB.cpp#L120-L159

## Sequence diagram
image

## Configuration options
```xml
  <syscheck>
   
   [...]
   
   <synchronization>
      <enabled>yes</enabled>
      <interval>5m</interval>
      <reponse_timeout>30</reponse_timeout>
      <max_interval>3600</max_interval>
      <max_eps>10</max_eps>
    </synchronization>
  </syscheck>
```

## Logs/Alerts example
Debug msg when the interval is doubled:
`DEBUG: Synchronization has failed, sync interval will be doubled, current interval: 600`

Debug msg when the interval is restored to the configured value:
`DEBUG: Interval synchronization has been reset to its original configured value: 300`

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
- [x] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities:
https://github.com/wazuh/wazuh-qa/issues/3053

- Memory tests for Linux
  - [x] Scan-build report
- Memory tests for Windows
  - [x] Scan-build report

- [x] Configuration on demand reports new parameters

### CODE REVIEW

**Review by antoniomanuelfr — APPROVED** [review]

File: src/config/syscheck-config.h [hunk 1]
```diff
@@ -403,8 +403,9 @@ typedef struct _config {
     char **nodiff;                                     /* list of files/dirs to never output diff */
     OSMatch **nodiff_regex;                            /* regex of files/dirs to never output diff */
 
-    long sync_interval;                                /* Synchronization interval (seconds) */
-    uint32_t min_sync_interval;                        /* Minimum interval for synchronization process */
+    uint32_t sync_interval;                            /* Synchronization interval (seconds) */
```
Comment (chemamartinez) [inline src/config/syscheck-config.h:406]:
```suggestion
    uint32_t sync_interval;                            /* Synchronization interval */
```

File: src/syscheckd/src/db/src/fimDB.cpp [hunk 2]
```diff
@@ -46,14 +45,51 @@ void FIMDB::sync()
...
+{
+    char debugmsg[1024];
+
+    if ((uint32_t)(getCurrentTime() - m_timeLastSyncMsg) > m_syncResponseTimeout)
+    {
+        if (m_syncSuccessful && m_currentSyncInterval > m_syncInterval)
+        {
+            m_currentSyncInterval = m_syncInterval;
+
+            snprintf(debugmsg, 1024, "FIM Sync check success, sync interval reset to original value: %ds", m_currentSyncInterval);
```
Comment (chemamartinez) [inline src/syscheckd/src/db/src/fimDB.cpp:58]:
```suggestion
            snprintf(debugmsg, 1024, "Previous sync was successful. Sync interval is reset to: '%ds'", m_currentSyncInterval);
```

File: src/syscheckd/src/db/src/fimDB.cpp [hunk 3]
```diff
@@ -46,14 +45,51 @@ void FIMDB::sync()
...
+    else
+    {
+        m_currentSyncInterval *= 2;
+
+        if (m_currentSyncInterval > m_syncMaxInterval)
+        {
+            m_currentSyncInterval = m_syncMaxInterval;
+        }
+
+        snprintf(debugmsg, 1024, "FIM Sync check failed, sync interval increased. Next interval: %ds", m_currentSyncInterval);
```
Comment (chemamartinez) [inline src/syscheckd/src/db/src/fimDB.cpp:74]:
```suggestion
        snprintf(debugmsg, 1024, "Sync still in progress. Skipped next sync and increased interval to '%ds'", m_currentSyncInterval);
```

File: src/syscheckd/src/db/tests/db/FIMDB/fimDBTests/fimDBImpTests.cpp [hunk 4]
```diff
@@ -252,6 +272,31 @@ TEST_F(FimDBFixture, loopRSyncSuccess)
     fimDBMock.runIntegrity();
 }
 
+TEST_F(FimDBFixture, syncAlgorithmLoop)
+{
+    EXPECT_CALL(fimDBMock, getCurrentTime()).WillOnce(testing::Return(15)).WillOnce(testing::Return(30));
+    EXPECT_CALL(*mockLog, loggingFunction(LOG_DEBUG_VERBOSE, "FIM Sync check failed, sync interval increased. Next interval: 1800s"));
```
Comment (chemamartinez) [inline src/syscheckd/src/db/tests/db/FIMDB/fimDBTests/fimDBImpTests.cpp:278]: Adapt the logs in the test with the suggestions from the code.

**Review by chemamartinez — CHANGES_REQUESTED** [review]

File: src/config/syscheck-config.c [hunk 5]
```diff
@@ -1198,10 +1198,22 @@ static void parse_synchronization(syscheck_config * syscheck, XML_NODE node) {
...
+            long max_interval = w_parse_time(node[i]->content);
+
+            if (max_interval < 0) {
+                mwarn(XML_VALUEERR, node[i]->element, node[i]->content);
+            } else {
+                syscheck->sync_max_interval = (uint32_t) max_interval;
+            }
         } else if (strcmp(node[i]->element, xml_response_timeout) == 0) {
-            mdebug1("'%s' has been deprecated. This setting is skipped.", xml_response_timeout);
+            long interval = w_parse_time(node[i]->content);
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply src/config/syscheck-config.c:1210]: Done

File: src/config/syscheck-config.c [hunk 6]
```diff
@@ -1223,21 +1235,15 @@ static void parse_synchronization(syscheck_config * syscheck, XML_NODE node) {
...
             mwarn(XML_INVELEM, node[i]->element);
         }
     }
 
-    if (syscheck->min_sync_interval >= syscheck->sync_interval) {
-        mwarn("Sync min_interval value higher than interval. Some synchronization turns can be skipped.");
+    if (syscheck->sync_max_interval < syscheck->sync_interval) {
+        syscheck->sync_max_interval = syscheck->sync_interval;
+
+        mwarn("Sync max_interval value cannot be less than interval. Setting max_interval to interval value: %d.", syscheck->sync_interval);
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply src/config/syscheck-config.c:1246]: Done

File: src/config/syscheck-config.h [hunk 7]
```diff
@@ -403,8 +403,9 @@ typedef struct _config {
...
+    uint32_t sync_response_timeout;                    /* Minimum interval for synchronization process */
+    uint32_t sync_max_interval;                        /* Maximun interval allowed for synchronization process */
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply src/config/syscheck-config.h]: Done

File: src/syscheckd/src/db/include/db.h [hunk 8]
```diff
@@ -33,21 +33,23 @@ extern "C" {
  * It will be dbsync the responsible of managing the DB.
  * @param storage storage 1 Store database in memory, disk otherwise.
  * @param sync_interval Interval when the synchronization will be performed.
+ * @param sync_max_interval Maximun interval allowed for synchronization process.
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply src/syscheckd/src/db/include/db.h]: Done

File: src/syscheckd/src/db/include/db.h [hunk 9]
```diff
@@ -33,21 +33,23 @@ extern "C" {
...
+ * @param sync_response_timeout Minimum interval for synchronization process.
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply src/syscheckd/src/db/include/db.h:37]: Done

File: src/config/syscheck-config.h (hunk 1 above)
Comment (jotacarma90) [reply src/config/syscheck-config.h:406]: Done

File: src/syscheckd/src/db/src/fimDB.cpp (hunk 2 above)
Comment (jotacarma90) [reply src/syscheckd/src/db/src/fimDB.cpp:58]: Done

File: src/syscheckd/src/db/src/fimDB.cpp (hunk 3 above)
Comment (jotacarma90) [reply src/syscheckd/src/db/src/fimDB.cpp:74]: Done

**Review by antoniomanuelfr — APPROVED** [review]

File: src/config/syscheck-config.c [hunk 10]
```diff
@@ -1223,21 +1235,15 @@ static void parse_synchronization(syscheck_config * syscheck, XML_NODE node) {
...
-            }
         } else {
             mwarn(XML_INVELEM, node[i]->element);
         }
     }
 
-    if (syscheck->min_sync_interval >= syscheck->sync_interval) {
-        mwarn("Sync min_interval value higher than interval. Some synchronization turns can be skipped.");
+    if (syscheck->sync_max_interval < syscheck->sync_interval) {
+        syscheck->sync_max_interval = syscheck->sync_interval;
```
Comment (FrancoRivero) [inline src/config/syscheck-config.c:1244]: If this is a warning, why have you assigned a value to sync_max_interval?

Comment (chemamartinez) [reply src/config/syscheck-config.c:1244]: It is a warning to alert the user that something in the configuration is wrong. However, we can set the closest value to the user choice and continue with the program.

**Review by FrancoRivero — APPROVED** [review]
LGTM

**Pull request merged**